lib/models/package-info-cache/package-info.js

```diff
@@ -415,6 +490,126 @@ class PackageInfo {
 
     return (this.addonConstructor = ctor);
   }
+
+  
+  /**
+   * Construct an addon instance.
+   *
+   * NOTE: this does NOT call constructors for the child addons. That is left to
+   * the caller to do, so they can insert any other logic they want.
+   *
+   * XXX THE ABOVE IS WRONG! The CoreObject init() code calls setupRegistry(),
+   * which in turn instantiates the addons. That defeats what we are trying to do here.
+   * @private
+   * @method constructAddonInstance
+   * @param {Project|Addon} parent the parent that directly contains this addon
+   * @param {Project} project the project that is/contains this addon
+   */
+  constructAddonInstance(parent, project) {
+    let start = Date.now();
+
+    let AddonConstructor = this.getAddonConstructor();
+
+    let addon;
+
+    try {
+      addon = new AddonConstructor(parent, project);
+    } catch (e) {
+      if (parent && parent.ui) {
+        parent.ui.writeError(e);
+      }
+      const SilentError = require('silent-error');
+      throw new SilentError(`An error occurred in the constructor for ${this.name} at ${this.realPath}`);
+    }
+
+    AddonConstructor._meta_.initializeIn = Date.now() - start;
+    addon.constructor = AddonConstructor;
+
+    return addon;
+  }
+
+  /**
+   * Create an instance of the addon represented by this packageInfo or (if we
+   * are supporting singletons and this is a singleton-enabled addon) check if
+   * we should be creating a proxy instead.
+   * NOTE: we assume that the value of 'isSingleton' does not change between
+   * calls to the constructor! A given addon is either a singleton or not for an
+   * entire run.
+   *
+   * @method getAddonInstance
+   * @param {} parent the addon/project that is to be the direct parent of the
+   * addon instance created here
+   * @param {*} project the project that is to contain this addon instance
+   * @return {Object} the constructed instance of the addon
+   */
+  getAddonInstance(parent, project) {
+    let ctor = this.getAddonConstructor();
+    let addon;
+
+    if (!SUPPORT_ADDON_INSTANCE_CACHING || !ctor.allowCachingPerBundle) {
```

Comment:
will the `ctor.allowCachingPerBundle` conditional always work? eg, if my addon entry point looks something like:

```js
module.exports = {
  name: 'foo',
  allowCachingPerBundle: true
};
```
ultimately this will create a constructor that extends from `lib/models/addon.js`, but `ctor.allowCachingPerBundle` will be `undefined` AFAICT; we should also check `ctor.prototype.allowCachingPerBundle`

Reply:
I need to check this. Will put the logic into a utility function that checks the prototype, too, like 'is-lazy-engine.js'.

Reply:
Fixed to check the prototype.